src/test/java/de/tum/in/www1/artemis/service/connectors/vcs/VcsTokenRenewalServiceTest.java

```diff
@@ -0,0 +1,181 @@
+        ReflectionTestUtils.setField(getCurrentTokenManagementService(), "versionControlAccessToken", false);
+        ReflectionTestUtils.setField(vcsTokenRenewalService, "versionControlAccessToken", false);
+        gitlabRequestMockProvider.reset();
+    }
+
+    private VcsTokenManagementService getCurrentTokenManagementService() {
+        Object attribute = ReflectionTestUtils.getField(vcsTokenRenewalService, "vcsTokenManagementService");
+        if (attribute instanceof Optional<?> attributeOptional && attributeOptional.isPresent() && attributeOptional.get() instanceof VcsTokenManagementService service) {
+            return service;
+        }
+        else {
+            throw new Error("Attribute vcsTokenManagementService of VcsTokenRenewalService has wrong type");
+        }
+    }
+
+    private record UserData(String initialToken, Long initialLifetimeDays, String updatedToken) {
+
+        public UserData(String initialToken, Long initialLifetimeDays) {
+            this(initialToken, initialLifetimeDays, null);
+        }
+
+        public boolean tokenRenewalNecessary() {
+            return initialToken != null && updatedToken != null;
+        }
+
+        public boolean tokenCreationNecessary() {
+            return initialToken == null;
+        }
+    }
+
+    static Stream<? extends Arguments> userDataSource() {
+        return Stream.of(Arguments
+                .of(List.of(new UserData("uishfi", 234L), new UserData("sdfhsehfe", 2L, "oshdf"), new UserData(null, null, "sdhfihs"), new UserData("ofg4958", 27L, "e9h4th"),
+                        new UserData("fduvhid", 29L), new UserData("e9tertr", 364L), new UserData("fduvhid", -64L, "sdhfisf"), new UserData(null, null, "rhehofhs"))));
+    }
+
+    @ParameterizedTest
+    @MethodSource("userDataSource")
+    @WithMockUser(username = TEST_PREFIX + "student1", roles = "USER")
+    void testRenewAllPersonalAccessTokens(List<UserData> userData) throws GitLabApiException {
+        final Predicate<User> isUserOfTest = user -> user.getLogin().startsWith(TEST_PREFIX);
+
+        final List<User> users = userRepository.findAll().stream().filter(isUserOfTest).toList();
+        assertThat(userData.size()).isEqualTo(users.size());
+
+        final ArrayList<org.gitlab4j.api.models.User> gitlabUsers = new ArrayList<>();
+
+        for (int i = 0; i < users.size(); ++i) {
+            User user = users.get(i);
+            UserData data = userData.get(i);
+
+            user.setVcsAccessToken(data.initialToken());
+            if (data.initialLifetimeDays() != null) {
+                user.setVcsAccessTokenExpiryDate(ZonedDateTime.now().plusDays(data.initialLifetimeDays()));
+            }
+            else {
+                user.setVcsAccessTokenExpiryDate(null);
+            }
+            userRepository.save(user);
+
+            gitlabUsers.add(new org.gitlab4j.api.models.User().withId(user.getId()).withUsername(user.getLogin()));
+        }
```

Comment:
This test method is quite long. Can you extract some parts into helper methods? Currently it is hard to see where the setup ends, where the actual method under test is called, and then where the assertions happen (up to line 150, line 151, and lines 152 onward).

E.g. the marked part into a `setupGitLabUsers(users)`.